File: app/src/main/java/ru/art2000/extensions/activities/DayNightActivity.java
package ru.art2000.extensions.activities;

import android.annotation.SuppressLint;
import android.content.res.Configuration;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;

import java.util.Calendar;

import ru.art2000.calculator.R;
import ru.art2000.helpers.PrefsHelper;

@SuppressWarnings("RedundantSuppression")
@SuppressLint("Registered")
public class DayNightActivity extends AppCompatActivity {

    @Override
    protected void onResume() {
        super.onResume();
        if (isResumeNightModeChangeEnabled()) {
            requestThemeCheck();
        }
    }

    @Override
    public void setTheme(int resId) {
        int newMode;
        if (resId == R.style.RT_AppTheme_DayNight && isSetThemeNightModeChangeEnabled()) {
            if (nightModeCondition()) {
                newMode = AppCompatDelegate.MODE_NIGHT_YES;
            } else {
                newMode = AppCompatDelegate.MODE_NIGHT_NO;
            }
            if (PrefsHelper.isAppAutoDarkThemeBlack())
                resId = R.style.RT_AppTheme_DayNightBlack;
        } else if (resId == R.style.RT_AppTheme_System) {
            newMode = AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM;
            if (PrefsHelper.isAppAutoDarkThemeBlack())
                resId = R.style.RT_AppTheme_SystemBlack;
        } else if (resId == R.style.RT_AppTheme_Battery) {
            newMode = AppCompatDelegate.MODE_NIGHT_AUTO_BATTERY;
            if (PrefsHelper.isAppAutoDarkThemeBlack())
                resId = R.style.RT_AppTheme_BatteryBlack;
        } else if (resId == R.style.RT_AppTheme_Dark || resId == R.style.RT_AppTheme_Black) {
            newMode = AppCompatDelegate.MODE_NIGHT_YES;
        } else {
            newMode = AppCompatDelegate.MODE_NIGHT_NO;
        }

        if (AppCompatDelegate.getDefaultNightMode() != newMode) {
            onSetThemeNightModeChanged(newMode);
            AppCompatDelegate.setDefaultNightMode(newMode);
        }

        super.setTheme(resId);
    }

    @SuppressWarnings("unused")
    protected void onResumeNightModeChanged(int mode) {

    }

    protected boolean isResumeNightModeChangeEnabled() {
        return true;
    }

    @SuppressWarnings("unused")
    protected void onSetThemeNightModeChanged(int mode) {

    }

    protected boolean isSetThemeNightModeChangeEnabled() {
        return true;
    }

    public boolean nightModeCondition() {
        Calendar calendar = Calendar.getInstance();
        int hour = calendar.get(Calendar.HOUR_OF_DAY);
        int minute = calendar.get(Calendar.MINUTE);
        int second = calendar.get(Calendar.SECOND);

        int seconds = (hour * 60 + minute) * 60 + second;
        int startSeconds = PrefsHelper.getDarkThemeActivationTime();
        int endSeconds = PrefsHelper.getDarkThemeDeactivationTime();

        if (startSeconds < endSeconds) {
            return seconds >= startSeconds && seconds < endSeconds;
        } else {
            return seconds >= startSeconds || seconds < endSeconds;
        }
    }

    public boolean isDarkThemeApplied() {
        int nightModeFlags =
                getResources().getConfiguration().uiMode & Configuration.UI_MODE_NIGHT_MASK;

        return nightModeFlags == Configuration.UI_MODE_NIGHT_YES;
    }

    protected void requestThemeCheck() {
        int theme = PrefsHelper.getAppTheme();
        if (PrefsHelper.areDynamicColorsEnabled()) {
            theme = R.style.RT_AppTheme_System;
        }

        if (theme == R.style.RT_AppTheme_DayNight) {
            int newMode;

            if (nightModeCondition()) {
                newMode = AppCompatDelegate.MODE_NIGHT_YES;
            } else {
                newMode = AppCompatDelegate.MODE_NIGHT_NO;
            }

            if (AppCompatDelegate.getDefaultNightMode() != newMode) {
                onResumeNightModeChanged(newMode);
                AppCompatDelegate.setDefaultNightMode(newMode);
            }
        } else if (theme == R.style.RT_AppTheme_System) {
            int newMode = AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM;
            if (AppCompatDelegate.getDefaultNightMode() != newMode) {
                onResumeNightModeChanged(newMode);
                AppCompatDelegate.setDefaultNightMode(newMode);
            }
        } else if (theme == R.style.RT_AppTheme_Battery) {
            int newMode = AppCompatDelegate.MODE_NIGHT_AUTO_BATTERY;
            if (AppCompatDelegate.getDefaultNightMode() != newMode) {
                onResumeNightModeChanged(newMode);
                AppCompatDelegate.setDefaultNightMode(newMode);
            }
        }
    }
}

File: app/src/main/java/ru/art2000/extensions/activities/DayNightActivity.kt
package ru.art2000.extensions.activities

import androidx.appcompat.app.AppCompatActivity
import ru.art2000.calculator.R
import androidx.appcompat.app.AppCompatDelegate
import dagger.hilt.EntryPoint
import dagger.hilt.InstallIn
import dagger.hilt.android.AndroidEntryPoint
import dagger.hilt.android.EntryPointAccessors
import dagger.hilt.components.SingletonComponent
import ru.art2000.helpers.GeneralPreferenceHelper
import java.util.*

@Suppress("MemberVisibilityCanBePrivate")
@AndroidEntryPoint
open class DayNightActivity : AppCompatActivity() {

    val generalPrefsHelper: GeneralPreferenceHelper by lazy {
        val holder = EntryPointAccessors.fromApplication<PreferenceHelperHolder>(this)
        holder.prefsHelper
    }

    @EntryPoint
    @InstallIn(SingletonComponent::class)
    interface PreferenceHelperHolder {
        val prefsHelper: GeneralPreferenceHelper
    }

    override fun onResume() {
        super.onResume()
        if (isResumeNightModeChangeEnabled) {
            requestThemeCheck()
        }
    }

    override fun setTheme(resId: Int) {
        var actualResId = resId
        if (actualResId == R.style.RT_AppTheme_DayNight && isSetThemeNightModeChangeEnabled) {
            if (generalPrefsHelper.isAppAutoDarkThemeBlack) actualResId = R.style.RT_AppTheme_DayNightBlack
        } else if (actualResId == R.style.RT_AppTheme_System) {
            if (generalPrefsHelper.isAppAutoDarkThemeBlack) actualResId = R.style.RT_AppTheme_SystemBlack
        } else if (actualResId == R.style.RT_AppTheme_Battery) {
            if (generalPrefsHelper.isAppAutoDarkThemeBlack) actualResId = R.style.RT_AppTheme_BatteryBlack
        }

        super.setTheme(actualResId)
    }

    @Suppress("UNUSED_PARAMETER")
    protected fun onResumeNightModeChanged(@AppCompatDelegate.NightMode mode: Int) {}
    protected val isResumeNightModeChangeEnabled: Boolean
        get() = true

    @Suppress("UNUSED_PARAMETER")
    protected fun onSetThemeNightModeChanged(@AppCompatDelegate.NightMode mode: Int) {}
    protected val isSetThemeNightModeChangeEnabled: Boolean
        get() = true

    private fun nightModeCondition(): Boolean {
        val calendar = Calendar.getInstance()
        val hour = calendar[Calendar.HOUR_OF_DAY]
        val minute = calendar[Calendar.MINUTE]
        val second = calendar[Calendar.SECOND]
        val seconds = (hour * 60 + minute) * 60 + second
        val startSeconds = generalPrefsHelper.darkThemeActivationTime
        val endSeconds = generalPrefsHelper.darkThemeDeactivationTime
        return if (startSeconds < endSeconds) {
            seconds in startSeconds until endSeconds
        } else {
            seconds >= startSeconds || seconds < endSeconds
        }
    }

    protected fun requestThemeCheck() {
        var theme = generalPrefsHelper.appTheme
        if (generalPrefsHelper.areDynamicColorsEnabled) {
            theme = R.style.RT_AppTheme_System
        }
        if (theme == R.style.RT_AppTheme_DayNight) {
            val newMode = if (nightModeCondition()) {
                AppCompatDelegate.MODE_NIGHT_YES
            } else {
                AppCompatDelegate.MODE_NIGHT_NO
            }
            if (AppCompatDelegate.getDefaultNightMode() != newMode) {
                onResumeNightModeChanged(newMode)
                AppCompatDelegate.setDefaultNightMode(newMode)
            }
        } else if (theme == R.style.RT_AppTheme_System) {
            val newMode = AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM
            if (AppCompatDelegate.getDefaultNightMode() != newMode) {
                onResumeNightModeChanged(newMode)
                AppCompatDelegate.setDefaultNightMode(newMode)
            }
        } else if (theme == R.style.RT_AppTheme_Battery) {
            val newMode = AppCompatDelegate.MODE_NIGHT_AUTO_BATTERY
            if (AppCompatDelegate.getDefaultNightMode() != newMode) {
                onResumeNightModeChanged(newMode)
                AppCompatDelegate.setDefaultNightMode(newMode)
            }
        }
    }
}